src/Type/Php/ArraySumFunctionDynamicReturnTypeExtension.php

```diff
@@ -0,0 +1,43 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Reflection\ParametersAcceptorSelector;
+use PHPStan\Type\Accessory\NonEmptyArrayType;
+use PHPStan\Type\Constant\ConstantIntegerType;
+use PHPStan\Type\DynamicFunctionReturnTypeExtension;
+use PHPStan\Type\FloatType;
+use PHPStan\Type\IntegerType;
+use PHPStan\Type\Type;
+use PHPStan\Type\TypeCombinator;
+
+final class ArraySumFunctionDynamicReturnTypeExtension implements DynamicFunctionReturnTypeExtension
+{
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return $functionReflection->getName() === 'array_sum';
+	}
+
+	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): Type
+	{
+		if (!isset($functionCall->args[0])) {
+			return ParametersAcceptorSelector::selectSingle($functionReflection->getVariants())->getReturnType();
+		}
+
+		$arrayType = $scope->getType($functionCall->args[0]->value);
+		$itemType = $arrayType->getIterableValueType();
+		if (
+			$itemType instanceof IntegerType
+			|| $arrayType instanceof NonEmptyArrayType && $itemType instanceof FloatType
+		) {
+			return $itemType;
+		}
+
+		return ParametersAcceptorSelector::selectSingle($functionReflection->getVariants())->getReturnType();
```

Comment:
Instead of the rest of the function, you can do just `return TypeCombinator::intersect($itemType, new UnionType([new IntegerType(), new FloatType()]))`.